### ISSUE TITLE
proposal: custom commands v2

### ISSUE THREAD

**Issue by Pedro-Pessoa** [botlabs-gg/yagpdb#1337]
I propose we rewrite part of the custom commands code that involves some breaking changes, hence the name v2.

# Background
Iirc, custom commands were first introduced to the bot back in 2016. Of course a lot of it has changed ever since, but the "core" logic has not changed much. Most of it is due to preventing breaking changes. But also, some logics could be cleaned up due to more recent versions of go. 

I'd like this issue to be a place to debate how could we make CC v2, and what changes we should aim.

### Role funcs
I've made a PR similar to this (https://github.com/botlabs-gg/yagpdb/pull/883).
There are a lot of different role funcs. I suggest we only have `mentionRole`, `hasRole`, `giveRole`, `removeRole`, `getRole` and `setRoles`.

Right now we have, for example, `hasRoleName` and `hasRoleID`. Why don't use just simply have a `hasRole` func that take cares of the args?

We also have `hasRoleName`, `hasRoleID`, `targetHasRoleName`, `targetHasRoleID`. That's 4 different funcs that could be reduced to a single one: `hasRole`.
The user argument could be optional, if no user is provided, we target the current user. And the role arg would be "smartly" inferred as well. So the caller could use ID, name, role object, etc.

### HasPermissions func

Similarly to the role funcs clean up, we could have just `tmplHasPermissions` instead of `tmplHasPermissions` and `tmplTargetHasPermissions`

#### Change the type
Change the type to be a pointer. The benefit is to be able to use append directly into the slice without needing to reassign it.
```
This:
{{$x := cslice}}
{{$x = $x.Append 1}}

Would become:
{{$x := cslice}}
{{$x.Append 1}}
````

#### Reduce reflection usage
There is a LOT of unnecessary reflection usage in the Slice logics. This could be reduce drastically which would increase performance.

#### Allow to append/set multiple values at once
The append/set methods should take n args and append them all to the slice, instead of the user having to call append multiple times.

#### Create a Del method
Since the type is now a pointer, a Del method makes sense.

#### Allow the Set method to set multiple values at once
Similarly to the Slice proposal, allow the Set method to add multiple values at once, instead of having to call .Set multiple times.

#### Try catch
Since we now have try catch in the template package, all functions that could error should error, so that we can actually use try catch.

#### Standardize errors
There is no common pattern in the errors. Some have capital letters, some don't. I propose we standardize the error messages.

### Limits
Right now it is pretty hard to understand how limits work. 
Take this for example: https://github.com/botlabs-gg/yagpdb/blob/master/customcommands/tmplextensions.go#L746
The limit is set inside a function arbitrarily. 
This should be a top level constant, imo.

The `CheckGuildDBLimit` behaves similarly. 

Other than that, there are some limits that are set on a per function basis like `IncreaseCheckCallCounterPremium` which makes it really hard to follow. 

### tmplSort
Although I was the one that PRed this function. It turned out the API is not so good nor intuitive. I propose we revisit this implementation. I'm not so sure how we could better approach this func, but I like @jo3-l idea. I'll let he chime in on that.

### Remove some funcs from the templates package

I don't think funcs like `toInt` and `toString` belong to this package. In fact, a few other packages import the templates package for these funcs alone. I think they should be in their own package.

### Remove the slice func
Remove the slice func and use the template's default slice func.

### Overall
I feel there is a lot of unnecessary reflection usage that could be cleaned up. I haven't dived deep into it yet, but as soon as I do I'll give my ideas here.

----

These are just a few things I remembered now. I'm pretty sure there are a lot of other things that could be improved.

**Comment by jo3-l** [comment]
Thanks for raising this issue; I wholeheartedly agree with the general sentiment that there is much to improve if we ever decide to release a backward-incompatible update to custom commands. That said, I think it is worth discussing whether *any* major update of this scale should be considered at all. By its very nature, such an update would break countless custom commands and potentially infuriate users, who are now forced to understand the many potentially subtle differences and then spend time rewriting their commands. Even worse, as custom commands have few compile-time checks, it's possible that certain commands would keep running with subtle changes in behaviour. (I would expect a non-negligible number to abandon YAGPDB completely as a consequence.)

**Comment by Pedro-Pessoa** [comment]
> Thanks for raising this issue; I wholeheartedly agree with the general sentiment that there is much to improve if we ever decide to release a backward-incompatible update to custom commands. That said, I think it is worth discussing whether _any_ major update of this scale should be considered at all. By its very nature, such an update would break countless custom commands and potentially infuriate users, who are now forced to understand the many potentially subtle differences and then spend time rewriting their commands. Even worse, as custom commands have few compile-time checks, it's possible that certain commands would keep running with subtle changes in behaviour. (I would expect a non-negligible number to abandon YAGPDB completely as a consequence.)

It just became really troublesome to maintain the code and we also have a convoluted API.
If we want to maintain back comp (which I don't think we should, but that's up to debate) we could do the same as we did for the amod. Make the current implementation cc legacy and create the new cc system.

**Comment by Pedro-Pessoa** [comment]
I think, if @ashishjh-bst agrees with this, he could open a cc2 (or something similar) branch for us, and we could start opening PRs to that branch and discuss each change individually. 

## Slices
I'm not in favor of adding more APIs. I think changing the behavior of the current methods is best. 
Adding more methods has no benefit but to maintain back comp, and I stated earlier my points on back comp.

## Errors
I agree!

## Sorting
Sounds good to me.

## Limits
I have to disagree. If I tell someone to find where the limitation of the DB value size is set, I bet they'd have a hard time if they didn't know it beforehand. I don't think this is good. 
Having it be limited by a "magic number" inside a func body is not ideal imo.
Even a func that you PRed goes in a different route: https://github.com/botlabs-gg/yagpdb/blob/963c5a6db701791e605cbc3524f84413f6034fc0/common/templates/context_funcs.go#L789
The role limit is set in a const outside the function body. I think this makes much more sense.

## Reflection
I'd like to see your points as to why, for example, we need reflection here: https://github.com/botlabs-gg/yagpdb/blob/963c5a6db701791e605cbc3524f84413f6034fc0/common/templates/context.go#L831

**Comment by Pedro-Pessoa** [comment]
Another thing I find a bit odd and I'd like to bring to the debate is whether the `templates` package belong inside common. 
What if it was a sub package of the custom commands package? Wouldn't it make more sense?

**Comment by l-zeuch** [comment]
To add, even though this is less of a breaking change: @jo3-l suggested a while back in our private channels to think about compiling custom command code to its bytecode and storing that.

I think this could be a somewhat great opportunity to implement that, as we're already reworking basically the entire logic. (Just saying it here for "all things considered").

Now I'll follow with my thoughts:

### Role functions et al

I definitely agree with reviewing and merging templates/functions that do the same thing just with different types. This'll make it easier to read custom command code as well, which also has to be considered.

Furthermore, this reduces the cognitive overhead for new users: as of now, I'd imagine they feel quite overwhelmed with what is essentially just a duplicate -- I, for one, definitely was.

I'll leave out sections where I agree and have nothing to add, to keep it somewhat short and simple.

### Slices

I'd prefer to have it as @Pedro-Pessoa suggested: Most new users that I've seen get bitten by this, assuming it is in-place whilst it is not. Myself I get bitten by it sometimes as well.

I disagree with adding yet another API `*InPlace`. This just adds duplicates and I cannot think of a reasonable use-case for it to not be in-place apart from constructed scenarios.

#### Append/Set multiple values at once

I am tending towards -1 on this one: on the one hand, it's sure nice to have, but I fear it'll introduce a lot of cognitive overhead from very long `.Set` calls. Personally, I believe it is better to leave it as-is.

### Errors

#### Standardize Errors

A consistent format throughout all functions is definitely favourable. Regarding the implementation of this, what are your thoughts on designating a file with a list of all errors we might return to the user, as well as assigning them a specific error code (similar to how Vim/Neovim do it)? I quite like the idea, but can see some caveats, such as:

* Error codes themselves might be obscure, so we have to document them.
* What if we add new errors? Do we have a system like the HTTP standard, where a certain range tells you the kind, and the rest tells you what specifically went wrong? Or is it incremental?

These are the most obvious ones that came to mind as of writing -- feel free to expand.

### Limits

I disagree with @jo3-l here: even though it might make some sense to have such constants/magic numbers close to the definition, having these constants (and some more) in one place makes hacking on the code way easier. 

That way, I don't have to search the codebase for any constants I might change, I just have them all in one place. This also makes it easier to add new constants, in my opinion.

----

> I think it is worth discussing whether any major update of this scale should be considered at all. By its very nature, such an update would break countless custom commands and potentially infuriate users, who are now forced to understand the many potentially subtle differences and then spend time rewriting their commands.

It might be worthwhile to split such one huge breaking change into smaller, more digestible small hunks. This way, users can just "fix" one breaking change at a time, keeping frustration hopefully low.

However, knowing the human nature, there will always be complaints. This is inevitable. To ease this, we should also consider the time between releasing a single breaking change in this context.

I understand that a decent amount of our users have the "setup and forget about it" mindset when it comes to custom commands -- if we consider the above points, hopefully this should level off any major issues and complaints.

>  Even worse, as custom commands have few compile-time checks, it's possible that certain commands would keep running with subtle changes in behaviour.

Just curious: what exactly do you mean by that? Is it the checks imposed by the Go compiler? If so, then we should use this opportunity to write tests for this part of the codebase.

----

> Another thing I find a bit odd and I'd like to bring to the debate is whether the templates package belong inside common.
> What if it was a sub package of the custom commands package? Wouldn't it make more sense?

Don't some partials for the frontend call functions from the template package? If so, perhaps we should consider moving those which are not called by these partials into `customcommands`.

Overall I agree with this though: templates for custom commands should reside in `customcommands`.

**Comment by ashishjh-bst** [comment]
If we go ahead with this, we cannot at all take this live without backwards compatibility, people spend hours writing CCs on yagpdb and have high dependency on their server functioning. 

We can at best add a deprecation notice with a defined period, after which the functions marked as deprecated can be removed. 

> 
Don't some partials for the frontend call functions from the template package? If so, perhaps we should consider moving those which are not called by these partials into

I see template as a lib like discordgo and goredis, so I'd prefer it stays in lib. There are more modules than custom commands using templates.

**Comment by Pedro-Pessoa** [comment]
> If we go ahead with this, we cannot at all take this live without backwards compatibility, people spend hours writing CCs on yagpdb and have high dependency on their server functioning.

There is no way we can make funcs return errors to take advantage of try catch that is backward compatible, afaik.

Of course this wouldn't be a "surprise" change. We'd need to make announcements that we are working on it, tell users what we intend to break, and finally release it in due time.
Or we could make a cc2 like an amod2, which I don't think is ideal, but is a possibility.

> I see template as a lib like discordgo and goredis, so I'd prefer it stays in lib. There are more modules than custom commands using templates.

I was referring to https://github.com/botlabs-gg/yagpdb/tree/master/common/templates
not https://github.com/botlabs-gg/yagpdb/tree/master/lib/template
